components/common/src/package_graph.rs

```diff
@@ -118,57 +151,79 @@ impl PackageGraph {
     /// Returns None if package was not in graph
     /// Returns Some(true) if package is removed
     fn do_remove(&mut self, package: &PackageIdent) -> Option<bool> {
-        match self.nodes.remove(&package) {
-            Some(idx) => {
-                // And remove from graph
-                match self.graph.remove_node(idx) {
-                    Some(ident) => assert_eq!(&ident, package),
-                    None => panic!(
-                        "removed node from map but it wasn't in the graph: {}",
-                        package
-                    ),
-                }
-                Some(true)
+        self.nodes.remove_by_left(&package).map(|(_, idx)| {
+            match self.graph.remove_node(idx) {
+                Some(ident) => assert_eq!(&ident, package),
+                None => panic!(
+                    "removed node from map but it wasn't in the graph: {}",
+                    package
+                ),
             }
-            None => None,
-        }
+            true
+        })
     }
 
     /// does a specific PackageIdent appear in the graph
     ///
     pub fn has_package(&self, package: &PackageIdent) -> bool {
-        self.nodes.contains_key(package)
+        self.nodes.contains_left(package)
     }
 
     fn count_edges(&self, package: &PackageIdent, direction: petgraph::Direction) -> Option<usize> {
-        match self.nodes.get(package) {
-            Some(&idx) => {
-                let deps: Vec<NodeIndex> = self.graph.neighbors_directed(idx, direction).collect();
-                Some(deps.len())
-            }
-            None => None,
-        }
+        self.nodes
+            .get_by_left(package)
+            .map(|&idx| self.graph.neighbors_directed(idx, direction).count())
     }
 
-    /// Returns the number of dependancies for a package
+    /// Returns the number of dependencies for a package
     ///
-    /// Returns None if the package is not in the graph
+    /// Returns `None` if the package is not in the graph
     pub fn count_deps(&self, ident: &PackageIdent) -> Option<usize> {
         self.count_edges(ident, petgraph::Outgoing)
     }
 
     /// Returns the number of package which have this package as a dependency
     ///
-    /// Returns None if the package is not in the graph
+    /// Returns `None` if the package is not in the graph
     pub fn count_rdeps(&self, ident: &PackageIdent) -> Option<usize> {
         self.count_edges(ident, petgraph::Incoming)
     }
 
+    fn neighbours(
+        &self,
+        package: &PackageIdent,
+        direction: petgraph::Direction,
+    ) -> Option<Vec<PackageIdent>> {
+        self.nodes.get_by_left(&package).map(|&idx| {
+            self.graph
+                .neighbors_directed(idx, direction)
+                .map(|n| self.nodes.get_by_right(&n))
+                .flat_map(|e| e.cloned()) // implied unwrap here is ok as we have consistency between `self.graph` and `self.nodes`
```

Comment:
I think `flat_map` makes a lot of sense when the closure returns a collection, but it can technically be used to take an iterator over `Option<T>` values and transform it into an iterator over `T` values due to the fact that `Option` implements the `IntoIter` trait. However, I think `filter_map` is the more idiomatic choice for this sort of situation.